book/PageParser.php

```diff
@@ -40,17 +40,20 @@ public function metadataIsSet( $id ) {
 	}
 
 	/**
-	 * return the list of the chapters with the summary if it exist.
+	 * Return the list of the chapters (based on the ws-summary HTML if it exists, otherwise via all internal links).
+	 * @param string[] $pageList Array of page titles.
+	 * @param string[] $namespaces Array of all localized namespace names of the current wiki.
 	 * @return Page[]
-	 * TODO retrive only main namespace pages ?
 	 */
 	public function getChaptersList( $pageList, $namespaces ) {
 		$list = $this->xPath->query( '//*[@id="ws-summary" or contains(@class,"ws-summary")]/descendant::a[not(contains(@href,"action=edit") or contains(@class,"extiw") or contains(@class,"external") or contains(@class,"internal") or contains(@class,"image"))]' );
 		$chapters = [];
 		/** @var DOMElement $link */
 		foreach ( $list as $link ) {
-			$title = str_replace( ' ', '_', $link->getAttribute( 'title' ) );
+			// Extract the page title (including namespace) from the relative URL in the link.
+			$title = urldecode( substr( $link->getAttribute( 'href' ), strlen( '/wiki/' ) ) );
```

Comment:
It would be nice to also remove the fragment (after the first '#')

`explode( '#', $url )[0]` should do the job.

Reply:
Good point. I've changed it to use `parse_url()`, and added some tests.